File: 42-trapping-rain-water/42-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let left = 0;
    let right = height.length - 1;
    let leftmax = height[0]; 
    let rightmax = height[right];
    let ans = 0;
    
    while(left < right) {
        if(height[left] < height[right]) {
            left++;
            if(height[left] < leftmax) {
                ans += leftmax - height[left];
            } else {
                leftmax = height[left];
            }
        } else {
            right--;
            if(height[right] < rightmax) {
                ans += rightmax - height[right];
            } else {
                rightmax = height[right];
            }
        }
    }
    
    return ans;
};

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        vector<int> preMax(n, 0), postMax(n, 0);
        preMax[0] = height[0];
        postMax[n-1] = height[n-1];
        for(int i = 1; i < n; i++) {
            preMax[i] = max(preMax[i-1], height[i]);
            postMax[n-1-i] = max(postMax[n-1-i+1], height[n-1-i]);
        }
        
        // for(int i = 0; i < n; i++) {
        //     cout << preMax[i] << " " << postMax[i] << "\n";
        // }
        
        int ans = 0;
        for(int i = 0; i < n; i++) {
            // cout << min(preMax[i], postMax[i]) << " " << height[i] << "\n";
            ans += min(preMax[i], postMax[i]) - height[i];
        }
        
        return ans;
    }
};